### PR TITLE
Implement auto'd optimal hashtable args and memory limitation checking

### PR DESCRIPTION
**Pull request by bocajnotnef**
Related to: #1050, #390, #1106, #732 

Issued in #1117

### CODE REVIEW

**Comment by bocajnotnef** [comment]
@ctb CR, please. I've yet to add tests/changelog.

**Comment by ctb** [comment]
cc @luizirber

On Mon, Jun 29, 2015 at 10:33:19AM -0700, Jake Fenton wrote:
> @ctb CR, please. I've yet to add tests/changelog. 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dib-lab/khmer/pull/1126#issuecomment-116769865
-- 
C. Titus Brown, ctbrown@ucdavis.edu

File: scripts/normalize-by-median.py [hunk 1]
```diff
@@ -251,6 +251,26 @@ def main():  # pylint: disable=too-many-branches,too-many-statements
...
+    # if optimization args are given, do optimization
+    if args.unique_kmers is not 0:
+        if args.max_memory_usage:
+            # verify that this is a sane memory usage restriction
+            res = oxutils.estimate_optimal_with_N_and_M(args.unique_kmers,
+                                                        args.max_memory_usage)
+            if res.fp_rate > 0.1:
+                print("""
+*** ERROR: The given restrictions yield an estimate false positive rate of {0},
+*** which is above the reccomended false positive ceiling of 0.1!
```
Comment (luizirber) [inline scripts/normalize-by-median.py]: typo

File: scripts/normalize-by-median.py [hunk 2]
```diff
@@ -251,6 +251,26 @@ def main():  # pylint: disable=too-many-branches,too-many-statements
     report_fp = args.report
     force_single = args.force_single
 
+    # if optimization args are given, do optimization
+    if args.unique_kmers is not 0:
```
Comment (luizirber) [inline scripts/normalize-by-median.py]: don't use `is` for comparison (exception: comparing to `True, False, None`)

File: khmer/khmer_args.py [hunk 3]
```diff
@@ -60,6 +60,8 @@ def build_hash_args(descr=None, epilog=None, parser=None):
     parser.add_argument('--n_tables', '-N', type=int,
                         default=DEFAULT_N_TABLES,
                         help='number of k-mer counting tables to use')
+    parser.add_argument('-U', '--unique-kmers', type=int, default=0,
+                        help='number of unique kmers in the input set')
```
Comment (mr-c) [inline khmer/khmer_args.py]: Isn't an estimate good enough?

**Comment by bocajnotnef** [comment]
- [ ] Is it mergeable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?
- [ ] Is the Copyright year up to date?

Everything looks good on Jenkins--some other change dropped test coverage (that is, `make diff-cover` comes up clean for me)

@ctb @luizirber CR/merge please?

File: scripts/normalize-by-median.py [hunk 4]
```diff
@@ -281,6 +284,10 @@ def main():  # pylint: disable=too-many-branches,too-many-statements
         print('loading k-mer counting table from ' + args.loadtable,
               file=sys.stderr)
         htable = khmer.load_counting_hash(args.loadtable)
+        if args.unique_kmers is not 0:
```
Comment (luizirber) [inline scripts/normalize-by-median.py]: `!=` instead of `is`

**Comment by bocajnotnef** [comment]
@luizirber Updated.

File: oxli/functions.py [hunk 5]
```diff
@@ -81,6 +83,37 @@ def optimal_args_output_gen(unique_kmers, fp_rate):
...
+        else:
+            res = estimate_optimal_with_N_and_f(args.unique_kmers,
+                                                desired_max_fp)
+            if args.max_tablesize and args.max_tablesize < res.htable_size:
+                print("*** Warning: The given tablesize is be too small!",
+                      file=sys.stderr)
+                print("*** Estimating false positive rate to be {0}".format(
+                      res.fp_rate), file=sys.stderr)
+            else:
+                print("*** INFO: set memory ceiling using atuo optimaztion.",
```
Comment (luizirber) [inline oxli/functions.py]: typo

**Comment by bocajnotnef** [comment]
@luizirber Aaand updated again. : D

File: oxli/functions.py [hunk 6]
```diff
@@ -81,6 +83,37 @@ def optimal_args_output_gen(unique_kmers, fp_rate):
...
+        else:
+            res = estimate_optimal_with_N_and_f(args.unique_kmers,
+                                                desired_max_fp)
+            if args.max_tablesize and args.max_tablesize < res.htable_size:
+                print("*** Warning: The given tablesize is be too small!",
+                      file=sys.stderr)
+                print("*** Estimating false positive rate to be {0}".format(
+                      res.fp_rate), file=sys.stderr)
+            else:
+                print("*** INFO: set memory ceiling using atuo optimization.",
```
Comment (luizirber) [inline oxli/functions.py]: Hmm, there were two typos, you only removed one =P

File: tests/test_normalize_by_median.py [hunk 7]
```diff
@@ -73,6 +73,51 @@ def test_normalize_by_median_unpaired_final_read():
...
+    infile = utils.get_temp_filename('test.fa')
+    in_dir = os.path.dirname(infile)
+
+    shutil.copyfile(utils.get_test_data('test-filter-abund-Ns.fq'), infile)
+
+    script = 'normalize-by-median.py'
+    args = ['-U', '83', infile]
+    (status, out, err) = utils.runscript(script, args, in_dir)
+
+    assert "*** INFO: set memory ceiling using atuo optimaztion." in err, err
```
Comment (luizirber) [inline tests/test_normalize_by_median.py]: And you need to update the tests too.

Comment (bocajnotnef) [inline tests/test_normalize_by_median.py]: gorramnit. I've read that sentence like twenty times now and *now* I see
"atuo"

Sigh.

File: oxli/functions.py [hunk 8]
```diff
@@ -81,6 +83,37 @@ def optimal_args_output_gen(unique_kmers, fp_rate):
...
+    # if optimization args are given, do optimization
+    if args.unique_kmers != 0:
+        if args.max_memory_usage:
+            # verify that this is a sane memory usage restriction
+            res = estimate_optimal_with_N_and_M(args.unique_kmers,
+                                                args.max_memory_usage)
+            if res.fp_rate > desired_max_fp:
+                print("""
+*** ERROR: The given restrictions yield an estimate false positive rate of {0},
+*** which is above the recommended false positive ceiling of 0.1!
```
Comment (luizirber) [inline oxli/functions.py]: Shouldn't this be `desired_max_fp`, instead of `0.1`?

File: oxli/functions.py [hunk 9]
```diff
@@ -81,6 +83,37 @@ def optimal_args_output_gen(unique_kmers, fp_rate):
...
+*** Aborting!""".format(res.fp_rate), file=sys.stderr)
+                sys.exit(1)
+        else:
+            res = estimate_optimal_with_N_and_f(args.unique_kmers,
+                                                desired_max_fp)
+            if args.max_tablesize and args.max_tablesize < res.htable_size:
+                print("*** Warning: The given tablesize is be too small!",
```
Comment (luizirber) [inline oxli/functions.py]: `is be to small`

**Comment by luizirber** [comment]
LGTM

File: oxli/functions.py [hunk 10]
```diff
@@ -81,6 +83,37 @@ def optimal_args_output_gen(unique_kmers, fp_rate):
...
 
+def do_sanity_checking(args, desired_max_fp):
+    # if optimization args are given, do optimization
+    if args.unique_kmers != 0:
+        if args.max_memory_usage:
+            # verify that this is a sane memory usage restriction
+            res = estimate_optimal_with_N_and_M(args.unique_kmers,
+                                                args.max_memory_usage)
+            if res.fp_rate > desired_max_fp:
+                print("""
```
Comment (mr-c) [inline oxli/functions.py]: No. It must be possible to get around this check with '--force'

**Comment by mr-c** [comment]
"Auto optimization" is too strong of a description of this process.

On Thu, Jul 9, 2015, 22:12 Luiz Irber <notifications@github.com> wrote:

> Merged #1126 <https://github.com/dib-lab/khmer/pull/1126>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dib-lab/khmer/pull/1126#event-352335847>.
>
-- 
Michael R. Crusoe: Programmer & Bioinformatician crusoe@ucdavis.edu
The lab for Data Intensive Biology; University of California, Davis
https://impactstory.org/MichaelRCrusoe http://twitter.com/biocrusoe

**Comment by bocajnotnef** [comment]
@mr-c I'm open to suggestions

File: scripts/normalize-by-median.py [hunk 11]
```diff
@@ -281,6 +284,10 @@ def main():  # pylint: disable=too-many-branches,too-many-statements
         print('loading k-mer counting table from ' + args.loadtable,
               file=sys.stderr)
         htable = khmer.load_counting_hash(args.loadtable)
+        if args.unique_kmers != 0:
+            print('Warning: You have specified a number of unique kmers'
+                  ' but are loading a precreated counting table--'
+                  'argument optimization will NOT be done.', file=sys.stderr)
```
Comment (ctb) [inline scripts/normalize-by-median.py]: Is this what you were concerned about @mr-c? We could change this to

`counting table size will NOT be set automatically.`

Comment (mr-c) [inline scripts/normalize-by-median.py]: +1

On Tue, Jul 14, 2015, 07:00 C. Titus Brown <notifications@github.com> wrote:

> In scripts/normalize-by-median.py
> <https://github.com/dib-lab/khmer/pull/1126#discussion_r34570329>:
>
> > @@ -281,6 +284,10 @@ def main():  # pylint: disable=too-many-branches,too-many-statements
> >          print('loading k-mer counting table from ' + args.loadtable,
> >                file=sys.stderr)
> >          htable = khmer.load_counting_hash(args.
>
> loadtable)
> > +        if args.unique_kmers != 0:
> > +            print('Warning: You have specified a number of unique kmers'
> > +                  ' but are loading a precreated counting table--'
> > +                  'argument optimization will NOT be done.', file=sys.stderr)
>
> Is this what you were concerned about @mr-c <https://github.com/mr-c>? We
> could change this to
>
> counting table size will NOT be set automatically.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dib-lab/khmer/pull/1126/files#r34570329>.
>
-- 
Michael R. Crusoe: Programmer & Bioinformatician crusoe@ucdavis.edu
The lab for Data Intensive Biology; University of California, Davis
https://impactstory.org/MichaelRCrusoe http://twitter.com/biocrusoe

**Pull request merged**